### PR TITLE
[Enhancement] Add back pressure to LakeTabletChannel when too many memtable need flush

### PR DESCRIPTION
**Pull request by meegoo**
backport https://github.com/StarRocks/starrocks/pull/22669 to LakeTabletChannel

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by sduzh** [comment]
@meegoo Could you please add some unit tests?

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 16 / 19 (84.21%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/async_delta_writer.cpp|1|3|33.33%|[89, 273]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|2|3|66.67%|[637]|
|:large_blue_circle:|src/runtime/lake_tablets_channel.cpp|13|13|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 12 / 20 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/lake_tablets_channel.cpp|3|11|27.27%|[310, 311, 312, 313, 314, 315, 317, 318]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|6|6|100.00%|[]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.cpp|3|3|100.00%|[]|

File: be/src/storage/lake/async_delta_writer.cpp [hunk 1]
```diff
@@ -53,6 +53,8 @@ class AsyncDeltaWriterImpl {
 
     void close();
 
+    [[nodiscard]] const int64_t queueing_memtable_num() const { return _writer->queueing_memtable_num(); }
```
Comment (sduzh) [inline be/src/storage/lake/async_delta_writer.cpp:56]:
```suggestion
    [[nodiscard]] int64_t queueing_memtable_num() const { return _writer->queueing_memtable_num(); }
```

File: be/src/storage/lake/async_delta_writer.cpp [hunk 2]
```diff
@@ -238,6 +240,10 @@ void AsyncDeltaWriter::close() {
     _impl->close();
 }
 
+const int64_t AsyncDeltaWriter::queueing_memtable_num() const {
```
Comment (sduzh) [inline be/src/storage/lake/async_delta_writer.cpp:243]:
```suggestion
int64_t AsyncDeltaWriter::queueing_memtable_num() const {
```

File: be/src/storage/lake/async_delta_writer.h [hunk 3]
```diff
@@ -86,6 +86,8 @@ class AsyncDeltaWriter {
     // [thread-safe]
     void close();
 
+    [[nodiscard]] const int64_t queueing_memtable_num() const;
```
Comment (sduzh) [inline be/src/storage/lake/async_delta_writer.h:89]:
```suggestion
    [[nodiscard]] int64_t queueing_memtable_num() const;
```

File: be/src/storage/lake/delta_writer.cpp [hunk 4]
```diff
@@ -114,6 +114,8 @@ class DeltaWriterImpl {
 
     [[nodiscard]] Status flush_async();
 
+    const int64_t queueing_memtable_num() const;
```
Comment (sduzh) [inline be/src/storage/lake/delta_writer.cpp:117]:
```suggestion
    int64_t queueing_memtable_num() const;
```

File: be/src/runtime/lake_tablets_channel.cpp [hunk 5]
```diff
@@ -302,6 +304,20 @@ void LakeTabletsChannel::add_chunk(Chunk* chunk, const PTabletWriterAddChunkRequ
...
+        while (dw->queueing_memtable_num() >= config::max_queueing_memtable_per_tablet) {
+            auto t1 = std::chrono::steady_clock::now();
+            if (std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count() / 1000 > request.timeout_ms()) {
+                LOG(INFO) << "LakeTabletsChannel txn_id: " << _txn_id << " load_id: " << print_id(request.id())
+                          << " wait tablet " << tablet_id << " flush memtable " << request.timeout_ms()
+                          << "ms still has queueing num " << dw->queueing_memtable_num();
+                break;
+            }
+            bthread_usleep(10000); // 10ms
+            wait_memtable_flush_time_us += 10000;
```
Comment (sduzh) [inline be/src/runtime/lake_tablets_channel.cpp:318]: bthread_usleep may sleep quite longer than expected when the bthread workers are busy

File: be/src/runtime/lake_tablets_channel.cpp [hunk 6]
```diff
@@ -302,6 +304,20 @@ void LakeTabletsChannel::add_chunk(Chunk* chunk, const PTabletWriterAddChunkRequ
         int64_t tablet_id = tablet_ids[row_indexes[from]];
         auto& dw = _delta_writers[tablet_id];
         DCHECK(dw != nullptr);
+
+        // back pressure OlapTableSink since there are too many memtables need to flush
+        while (dw->queueing_memtable_num() >= config::max_queueing_memtable_per_tablet) {
+            auto t1 = std::chrono::steady_clock::now();
+            if (std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count() / 1000 > request.timeout_ms()) {
```
Comment (sduzh) [inline be/src/runtime/lake_tablets_channel.cpp:311]:
```suggestion
            if (std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count() > request.timeout_ms()) {
```

File: be/src/runtime/lake_tablets_channel.cpp (hunk 5 above)
Comment (meegoo) [reply be/src/runtime/lake_tablets_channel.cpp:318]: any suggestion about this?
In addition, if the bthread worker is really full, then it does not seem to be a big problem if the rpc request is blocked for a longer time.

Comment (sduzh) [reply be/src/runtime/lake_tablets_channel.cpp:318]: How about `wait_memtable_flush_time_us += now() - t1`

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 12 / 20 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/lake_tablets_channel.cpp|3|11|27.27%|[310, 311, 312, 313, 314, 315, 317, 318]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|6|6|100.00%|[]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.cpp|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 12 / 20 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/lake_tablets_channel.cpp|3|11|27.27%|[310, 311, 312, 313, 314, 315, 317, 318]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|6|6|100.00%|[]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.cpp|3|3|100.00%|[]|

File: be/src/runtime/lake_tablets_channel.cpp (hunk 5 above)
Comment (meegoo) [reply be/src/runtime/lake_tablets_channel.cpp]: done, misunderstanding before.

**Review by sduzh — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 12 / 22 (54.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/lake_tablets_channel.cpp|3|13|23.08%|[310, 311, 312, 313, 314, 315, 317, 318, 319, 320]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|6|6|100.00%|[]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.cpp|3|3|100.00%|[]|

**Review by wyb — APPROVED** [review]

**Comment by sduzh** [comment]
@Mergifyio backport branch-3.1

**Comment by sduzh** [comment]
@Mergifyio backport branch-3.2

**Pull request merged**